subxt/src/blocks/extrinsic_types.rs

```diff
@@ -0,0 +1,888 @@
+    pub(crate) fn decode_from(
+        index: u32,
+        extrinsic_bytes: Arc<[u8]>,
+        client: C,
+        block_hash: T::Hash,
+        cached_events: CachedEvents<T>,
+        ids: ExtrinsicIds,
+    ) -> Result<ExtrinsicDetails<T, C>, Error> {
+        const SIGNATURE_MASK: u8 = 0b1000_0000;
+        const VERSION_MASK: u8 = 0b0111_1111;
+        const LATEST_EXTRINSIC_VERSION: u8 = 4;
+
+        let metadata = client.metadata();
+
+        // Extrinsic are encoded in memory in the following way:
+        //   - first byte: abbbbbbb (a = 0 for unsigned, 1 for signed, b = version)
+        //   - signature: [unknown TBD with metadata].
+        //   - extrinsic data
+        if extrinsic_bytes.is_empty() {
+            return Err(BlockError::InsufficientData.into());
+        }
+
+        let version = extrinsic_bytes[0] & VERSION_MASK;
+        if version != LATEST_EXTRINSIC_VERSION {
+            return Err(BlockError::UnsupportedVersion(version).into());
+        }
+
+        let is_signed = extrinsic_bytes[0] & SIGNATURE_MASK != 0;
+
+        // Skip over the first byte which denotes the version and signing.
+        let cursor = &mut &extrinsic_bytes[1..];
+
+        let mut address_start_idx = 0;
+        let mut address_end_idx = 0;
+
+        if is_signed {
+            address_start_idx = extrinsic_bytes.len() - cursor.len();
+
+            // Skip over the address, signature and extra fields.
+            scale_decode::visitor::decode_with_visitor(
+                cursor,
+                ids.address,
+                &metadata.runtime_metadata().types,
+                scale_decode::visitor::IgnoreVisitor,
+            )
+            .map_err(scale_decode::Error::from)?;
+            address_end_idx = extrinsic_bytes.len() - cursor.len();
+
+            scale_decode::visitor::decode_with_visitor(
+                cursor,
+                ids.signature,
+                &metadata.runtime_metadata().types,
+                scale_decode::visitor::IgnoreVisitor,
+            )
+            .map_err(scale_decode::Error::from)?;
+
+            scale_decode::visitor::decode_with_visitor(
+                cursor,
+                ids.extra,
+                &metadata.runtime_metadata().types,
+                scale_decode::visitor::IgnoreVisitor,
+            )
+            .map_err(scale_decode::Error::from)?;
+        }
+
+        let call_start_idx = extrinsic_bytes.len() - cursor.len();
+
+        // Ensure the provided bytes are sound.
+        scale_decode::visitor::decode_with_visitor(
+            &mut *cursor,
+            ids.call,
+            &metadata.runtime_metadata().types,
+            scale_decode::visitor::IgnoreVisitor,
+        )
+        .map_err(scale_decode::Error::from)?;
+
+        // Decode the pallet index, then the call variant.
+        let cursor = &mut &extrinsic_bytes[call_start_idx..];
+
+        if cursor.len() < 2 {
+            return Err(BlockError::InsufficientData.into());
+        }
```

Comment:
Aah yeah; in most places we just let `codec::Error` encapsulate this. Since we've decoded the whole thing above though, we must have more than 2 bytes and so I'd just remove this error check entirely and maybe a brief comment (that was we can remove that `InsufficientData` variant).

Your prev approach of decoding the bytes was probably safer; maybe I shouldn't have opened my mouth in the first place because that worked nicely even if a little more verbose :)

Reply:
Yep, that makes sense, we can utilize that codec::Error. Although I also like that we avoided some decodes which made the code a bit more efficient :D